src/ros_filter.cpp

```diff
@@ -1506,136 +1637,22 @@ namespace RobotLocalization
-             * from odomFrameId_->baseLinkFrameId_. Now imagine that we have a position in the
-             * mapFrameId_ frame. First, we multiply it by the inverse of the
-             * mapFrameId_->baseLinkFrameId, which will transform that data from mapFrameId_ to
-             * baseLinkFrameId_. Now we want to go from baseLinkFrameId_->odomFrameId_, but the
-             * transform we have takes data from odomFrameId_->baseLinkFrameId_, so we need its
-             * inverse as well. We have now transformed our data from mapFrameId_ to odomFrameId_.
-             * However, if we want other users to be able to do the same, we need to broadcast
-             * the inverse of that entire transform.
-            */
-
-            mapOdomTrans.mult(worldBaseLinkTrans, odomBaseLinkTrans);
-
-            mapOdomTransMsg.transform = tf2::toMsg(mapOdomTrans);
-            mapOdomTransMsg.header.stamp = filteredPosition.header.stamp + tfTimeOffset_;
-            mapOdomTransMsg.header.frame_id = mapFrameId_;
-            mapOdomTransMsg.child_frame_id = odomFrameId_;
-
-            worldTransformBroadcaster.sendTransform(mapOdomTransMsg);
-          }
-          catch(...)
-          {
-            ROS_ERROR_STREAM("Could not obtain transform from " << odomFrameId_ << "->" << baseLinkFrameId_);
-          }
-        }
-        else
-        {
-          ROS_ERROR_STREAM("Odometry message frame_id was " << filteredPosition.header.frame_id <<
-                           ", expected " << mapFrameId_ << " or " << odomFrameId_);
-        }
-
-        // Fire off the position and the transform
-        positionPub.publish(filteredPosition);
-
-        if (printDiagnostics_)
-        {
-          freqDiag.tick();
-        }
-      }
-
-      // The spin will call all the available callbacks and enqueue
-      // their received measurements
-      ros::spinOnce();
-
-      // Now we'll integrate any measurements we've received
-      ros::Time curTime = ros::Time::now();
-      integrateMeasurements(ros::Time::now().toSec());
-
-      /* Diagnostics can behave strangely when playing back from bag
-       * files and using simulated time, so we have to check for
-       * time suddenly moving backwards as well as the standard
-       * timeout criterion before publishing. */
-      double diagDuration = (curTime - lastDiagTime).toSec();
-      if (printDiagnostics_ && (diagDuration >= diagnosticUpdater_.getPeriod() || diagDuration < 0.0))
-      {
-        diagnosticUpdater_.force_update();
-        lastDiagTime = curTime;
-      }
-
-      if (!loop_rate.sleep())
-      {
-        ROS_WARN_STREAM("Failed to meet update rate! Try decreasing the rate, limiting "
-                        "sensor output frequency, or limiting the number of sensors.");
-      }
+        //Not sure how to figure out what's the wait duration, currently going with 5ms
+        ros::getGlobalCallbackQueue()->callAvailable(ros::WallDuration(0.005));
```

Comment:
@ayrton04  The callAvailable uses a timed_wait on a boost condition_variable. It would just wait 5ms for a callback to be available and if a callback is queued during that time, it will be processed right away. So this is not supposed to incur any delay at all.